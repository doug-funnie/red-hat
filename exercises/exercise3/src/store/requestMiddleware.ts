import { Middleware } from 'redux';
import { Action } from 'redux-actions';

export enum Lifecycle {
  requested,
  success,
  error,
}

interface RequestAction<SP = any, RP = undefined> {
  type: string;
  request(a: RP): Promise<SP>;
  payload: RP;
  meta?: any;
}

export interface RequestMiddlewareAction<P = any> {
  type: string;
  payload: P;
  meta: {
    'request-middleware': Lifecycle;
    [key: string]: any;
  };
  error?: boolean;
}

// forces typings to match
type RequestMiddlewareKey = keyof Pick<RequestMiddlewareAction['meta'], 'request-middleware'>;
export const requestMiddlewareKey: RequestMiddlewareKey = 'request-middleware';

const isRequestAction = (action: RequestAction): action is RequestAction => !!action.request;
const isRequestMiddlewareAction = (action: RequestMiddlewareAction): action is RequestMiddlewareAction => (
  action.meta && action.meta[requestMiddlewareKey] !== null
);

export function createRequestAction<SP>(type: string, request: () => Promise<SP>): () => RequestAction<SP>;
export function createRequestAction<RP, SP>(
  type: string,
  request: (requestPayload: RP) => Promise<SP>,
): (a: RP) => RequestAction<SP, RP>;
export function createRequestAction(type, request) {
  return (a): RequestAction => ({
    type,
    request,
    payload: a,
  });
}

const requestHasExpired = (lastRequestedTime: number) => {
  if (!lastRequestedTime) {
    return true;
  }
  const diff = performance.now() - lastRequestedTime;
  return diff >= 300000; // 5 Minutes
};

export const createRequestMiddleware = () => {
  const requestCache: {
    [url: string]: {
      request: Promise<any>;
      lastRequestedTime: number;
    };
  } = {};
  const cacheRequest = (type: string, request: Promise<any>) => {
    requestCache[type] = {
      request,
      lastRequestedTime: performance.now(),
    };
  };

  const middleware: Middleware = ({ dispatch }) => (next) => async (action) => {
    if (!isRequestAction(action)) {
      return next(action);
    }

    const cachedRequest = requestCache[action.type];
    if (cachedRequest && cachedRequest.request && !requestHasExpired(cachedRequest.lastRequestedTime)) {
      return cachedRequest.request;
    }

    const createRequestMiddlewareAction = (lifecycle: Lifecycle, payload: any): RequestMiddlewareAction => ({
      payload,
      type: action.type,
      error: lifecycle === Lifecycle.error,
      meta: {
        ...action.meta,
        [requestMiddlewareKey]: lifecycle,
      },
    });

    dispatch(createRequestMiddlewareAction(Lifecycle.requested, action.payload));
    try {
      const request = action.request(action.payload);
      cacheRequest(action.type, request);
      const data = await(request);
      dispatch(createRequestMiddlewareAction(Lifecycle.success, data));
      return data;
    } catch (err) {
      dispatch(createRequestMiddlewareAction(Lifecycle.error, err));
      return Promise.reject(err);
    }
  };

  return middleware;
};

type RequestActionHandler<S = any, P = any> = (prevState: S, action: Action<P>) => S;

interface RequestActionHandlers<S, RP, SP, EP> {
  requested?: RequestActionHandler<S, RP>;
  success?: RequestActionHandler<S, SP>;
  error?: RequestActionHandler<S, EP>;
  finally?: RequestActionHandler<S, RP | EP>;
}

export function handleRequestAction<S, SP, RP = any, EP = any>(
  type: string,
  handlers: RequestActionHandlers<S, RP, SP, EP>,
  defaultState: S,
) {
  return (state: S = defaultState, action: RequestMiddlewareAction) => {
    if (action.type !== type) {
      return state;
    }

    if (!isRequestMiddlewareAction(action)) {
      return state;
    }

    const lifecycle = action.meta[requestMiddlewareKey];
    const handler = handlers[Lifecycle[lifecycle]];
    if (handler) {
      return handler(state, action);
    }
    return state;
  };
}
