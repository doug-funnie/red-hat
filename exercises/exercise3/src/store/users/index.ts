export * from './users.actions';
export * from './users.reducers';
export * from './users.selectors';
