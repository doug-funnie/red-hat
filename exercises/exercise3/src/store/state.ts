import { MapDispatchToProps, MapStateToProps } from 'react-redux';
import { UserListState, UsersState } from './users';

export interface RootState {
  users: UsersState;
  userList: UserListState;
}

export const enum SortDirection {
  asc,
  desc,
}

export const enum FetchStatus {
  none = 'none',
  inProgress = 'inProgress',
  complete = 'complete',
}

const createStateKey = <K extends keyof RootState>(key: K): K => key;
export const usersStateKey = createStateKey('users');
export const userListStateKey = createStateKey('userList');

export type TypedMapStateToProps<InnerProps, OuterProps> = MapStateToProps<InnerProps, OuterProps, RootState>;
export type TypedMapDispatchToProps<DispatchProps> = MapDispatchToProps<DispatchProps, any>;
