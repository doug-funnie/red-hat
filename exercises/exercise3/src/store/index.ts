import { applyMiddleware, createStore } from 'redux';
import { combineReducers } from 'redux';
import thunk from 'redux-thunk';
import { createRequestMiddleware } from './requestMiddleware';
import { RootState, userListStateKey, usersStateKey } from './state';
import { userListReducer, usersReducer } from './users';

export const rootReducer = combineReducers<RootState>({
  [userListStateKey]: userListReducer,
  [usersStateKey]: usersReducer,
});

export function create() {
  return createStore(
    rootReducer,
    applyMiddleware(
      thunk,
      createRequestMiddleware(),
    ),
  );
}
