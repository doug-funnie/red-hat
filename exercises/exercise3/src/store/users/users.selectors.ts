import { User } from 'api/users';
import { createSelector } from 'reselect';
import { FetchStatus, RootState, SortDirection } from '../state';

export const selectUsersState = (state: RootState) => state.users;
export const selectUserById = (state: RootState, id: User['id']) => selectUsersState(state).byId[id];
export const selectUserError = (state: RootState, id: User['id']) => selectUsersState(state).errors[id];
export const selectUserFetchStatus = (state: RootState, id: User['id']) => (
  selectUsersState(state).fetchStatus[id] || FetchStatus.none
);

export const selectUserListState = (state: RootState) => state.userList;
export const selectUserListUsers = (state: RootState) => selectUserListState(state).users;
export const selectUserListFetchStatus = (state: RootState) => selectUserListState(state).fetchStatus;
export const selectUserListSortDirection = (state: RootState) => selectUserListState(state).sortDirection;
export const selectUserListSortKey = (state: RootState) => selectUserListState(state).sortKey;
export const selectUserList = createSelector(
  [selectUserListUsers, selectUsersState],
  (userIds, usersState) => userIds.map((uId) => usersState.byId[uId]),
);
export const selectUserListSorted = createSelector(
  [selectUserList, selectUserListSortDirection, selectUserListSortKey],
  (userList, sortDirection, sortKey) => [...userList].sort((a, b) => {
    const aValue = a[sortKey];
    const bValue = b[sortKey];
    const isAsc = sortDirection === SortDirection.asc;
    if (aValue > bValue) { return isAsc ? 1 : -1; }
    if (aValue < bValue) { return isAsc ? -1 : 1; }
    return 0;
  }),
);
