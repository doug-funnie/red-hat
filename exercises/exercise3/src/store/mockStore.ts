import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { createRequestMiddleware } from 'store/requestMiddleware';

export function mockStoreCreator<S>(reducer: any) {
  return () => createStore<S>(
    reducer,
    applyMiddleware(
      thunk,
      createRequestMiddleware(),
    ),
  );
}
