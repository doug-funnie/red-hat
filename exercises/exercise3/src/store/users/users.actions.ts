import { getUser, getUsers, User } from 'api/users';
import { createRequestAction } from 'store/requestMiddleware';

export const FETCH_USERS = 'FETCH_USERS';
export const FETCH_USER = 'FETCH_USER';

export const fetchUsers = createRequestAction(FETCH_USERS, getUsers);
export const fetchUser = createRequestAction(
  FETCH_USER,
  (id: User['id']) => getUser(id).catch((error) => Promise.reject({ id, error }),
));
