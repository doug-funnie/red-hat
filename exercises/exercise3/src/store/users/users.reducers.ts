import { User } from 'api/users';
import reduceReducers from 'reduce-reducers';
import { handleRequestAction } from 'store/requestMiddleware';
import { FetchStatus, SortDirection } from '../state';
import { FETCH_USER, FETCH_USERS } from './users.actions';

export interface UsersState {
  byId: { [key: number]: User; };
  errors: { [key: number]: any; };
  fetchStatus: { [key: number]: FetchStatus; };
}

export interface UserListState {
  error: any;
  sortDirection: SortDirection;
  sortKey: keyof User;
  users: User['id'][];
  fetchStatus: FetchStatus;
}

const defaultUsersState: UsersState = {
  byId: {},
  errors: {},
  fetchStatus: {},
};

const defaultUserListState: UserListState = {
  error: null,
  sortDirection: SortDirection.asc,
  sortKey: 'name',
  users: [],
  fetchStatus: FetchStatus.none,
};

export const usersReducer = reduceReducers(
  handleRequestAction<UsersState, User[]>(FETCH_USERS, {
    success: (state, { payload: users }) => users.reduce((acc, user) => ({
      ...acc,
      byId: {
        ...acc.byId,
        [user.id]: user,
      },
      fetchStatus: {
        ...acc.fetchStatus,
        [user.id]: FetchStatus.complete,
      },
    }), state),
  }, defaultUsersState),
  handleRequestAction<UsersState, User, User['id']>(FETCH_USER, {
    requested: (state, { payload: id }) => ({
      ...state,
      fetchStatus: {
        ...state.fetchStatus,
        [id]: FetchStatus.inProgress,
      },
    }),
    success: (state, { payload: user }) => ({
      ...state,
      byId: {
        ...state.byId,
        [user.id]: user,
      },
      fetchStatus: {
        ...state.fetchStatus,
        [user.id]: FetchStatus.complete,
      },
    }),
    error: (state, { payload: { id, error } }) => ({
      ...state,
      errors: {
        ...state.byId,
        [id]: error,
      },
      fetchStatus: {
        ...state.fetchStatus,
        [id]: FetchStatus.complete,
      },
    }),
  }, defaultUsersState),
);

export const userListReducer = handleRequestAction<UserListState, User[]>(FETCH_USERS, {
  requested: (state) => ({
    ...state,
    fetchStatus: FetchStatus.inProgress,
  }),
  success: (state, { payload: users }) => ({
    ...state,
    fetchStatus: FetchStatus.complete,
    users: users.map((u) => u.id),
  }),
  error: (state, { payload: error }) => ({
    ...state,
    error,
    fetchStatus: FetchStatus.complete,
  }),
}, defaultUserListState);
