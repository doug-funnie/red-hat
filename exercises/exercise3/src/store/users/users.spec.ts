jest.mock('api/users');

import { getUser, getUsers, User } from 'api/users';
import { Store } from 'react-redux';
import { combineReducers } from 'redux';
import { mockStoreCreator } from 'store/mockStore';
import { FetchStatus, RootState } from 'store/state';
import * as actions from './users.actions';
import * as reducers from './users.reducers';
import * as selectors from './users.selectors';

const user1: Partial<User> = {
  id: 1,
  name: 'Z User 1',
};
const user2: Partial<User> = {
  id: 2,
  name: 'A User2',
};

const mockGetUsers = getUsers as jest.Mock<{}>;
const mockGetUser = getUser as jest.Mock<{}>;

beforeEach(() => {
  mockGetUsers.mockReturnValue(Promise.resolve([user1, user2]));
  mockGetUser.mockReturnValue(Promise.resolve(user1));
});

describe('userList', () => {
  const storeCreator = mockStoreCreator<reducers.UserListState>(reducers.userListReducer);
  let store: Store<reducers.UserListState>;

  beforeEach(() => {
    store = storeCreator();
  });

  it('defaults state', () => {
    expect(store.getState()).toMatchSnapshot();
  });

  it('handles successful fetchUsers action', async () => {
    const promise = store.dispatch(actions.fetchUsers());
    expect(store.getState()).toMatchSnapshot();
    await promise;
    expect(store.getState()).toMatchSnapshot();
  });

  it('handles failed fetchUsers action', async () => {
    try {
      // Unhandled promise rejection without some sort of catch
      mockGetUsers.mockReturnValue(Promise.reject('Error'));
      await store.dispatch(actions.fetchUsers());
    } catch (_) { } //tslint:disable-line
    expect(store.getState()).toMatchSnapshot();
  });
});

describe('users', () => {
  const storeCreator = mockStoreCreator<reducers.UserListState>(reducers.usersReducer);
  let store: Store<reducers.UserListState>;

  beforeEach(() => {
    store = storeCreator();
  });

  it('defaults state', () => {
    expect(store.getState()).toMatchSnapshot();
  });

  it('handles successful fetchUsers action', async () => {
    await store.dispatch(actions.fetchUsers());
    expect(store.getState()).toMatchSnapshot();
  });

  it('handles successful fetchUser action', async () => {
    const promise =  store.dispatch(actions.fetchUser(user1.id));
    expect(store.getState()).toMatchSnapshot();
    await promise;
    expect(store.getState()).toMatchSnapshot();
  });

  it('handles failed fetchUser action', async () => {
    try {
      // Unhandled promise rejection without some sort of catch
      mockGetUser.mockReturnValue(Promise.reject('Error'));
      await store.dispatch(actions.fetchUser(user1.id));
    } catch (_) { } //tslint:disable-line
    expect(store.getState()).toMatchSnapshot();
  });
});

describe('selectors', () => {
  let store: Store<Pick<RootState, 'users' | 'userList'>>;
  const createStore = mockStoreCreator<Pick<RootState, 'users' | 'userList'>>(
    combineReducers({
      users: reducers.usersReducer,
      userList: reducers.userListReducer,
    }),
  );

  beforeEach(async () => {
    store = createStore();
    await store.dispatch(actions.fetchUsers());
  });

  it('selectUserFetchStatus defaults to none', () => {
    expect(selectors.selectUserFetchStatus(store.getState(), 1000)).toBe(FetchStatus.none);
  });

  it('selectUserListSorted sorts user list using sortKey in userList state', () => {
    expect(selectors.selectUserListSorted(store.getState())).toMatchSnapshot();
  });

  it('selectUserListSorted returns same result if state is unchanged', () => {
    const first = selectors.selectUserListSorted(store.getState());
    const second = selectors.selectUserListSorted(store.getState());
    expect(first).toBe(second);
  });
});
