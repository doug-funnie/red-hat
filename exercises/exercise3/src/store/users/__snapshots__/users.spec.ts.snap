// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors selectUserListSorted sorts user list using sortKey in userList state 1`] = `
Array [
  Object {
    "id": 2,
    "name": "A User2",
  },
  Object {
    "id": 1,
    "name": "Z User 1",
  },
]
`;

exports[`userList defaults state 1`] = `
Object {
  "error": null,
  "fetchStatus": "none",
  "sortDirection": 0,
  "sortKey": "name",
  "users": Array [],
}
`;

exports[`userList handles failed fetchUsers action 1`] = `
Object {
  "error": "Error",
  "fetchStatus": "complete",
  "sortDirection": 0,
  "sortKey": "name",
  "users": Array [],
}
`;

exports[`userList handles successful fetchUsers action 1`] = `
Object {
  "error": null,
  "fetchStatus": "inProgress",
  "sortDirection": 0,
  "sortKey": "name",
  "users": Array [],
}
`;

exports[`userList handles successful fetchUsers action 2`] = `
Object {
  "error": null,
  "fetchStatus": "complete",
  "sortDirection": 0,
  "sortKey": "name",
  "users": Array [
    1,
    2,
  ],
}
`;

exports[`users defaults state 1`] = `
Object {
  "byId": Object {},
  "errors": Object {},
  "fetchStatus": Object {},
}
`;

exports[`users handles failed fetchUser action 1`] = `
Object {
  "byId": Object {},
  "errors": Object {
    "1": "Error",
  },
  "fetchStatus": Object {
    "1": "complete",
  },
}
`;

exports[`users handles successful fetchUser action 1`] = `
Object {
  "byId": Object {},
  "errors": Object {},
  "fetchStatus": Object {
    "1": "inProgress",
  },
}
`;

exports[`users handles successful fetchUser action 2`] = `
Object {
  "byId": Object {
    "1": Object {
      "id": 1,
      "name": "Z User 1",
    },
  },
  "errors": Object {},
  "fetchStatus": Object {
    "1": "complete",
  },
}
`;

exports[`users handles successful fetchUsers action 1`] = `
Object {
  "byId": Object {
    "1": Object {
      "id": 1,
      "name": "Z User 1",
    },
    "2": Object {
      "id": 2,
      "name": "A User2",
    },
  },
  "errors": Object {},
  "fetchStatus": Object {
    "1": "complete",
    "2": "complete",
  },
}
`;
